distributed/worker.py

```diff
@@ -3467,78 +3467,47 @@ def handle_worker_status_change(self, status: str, stimulus_id: str) -> None:
             # Update status and send confirmation to the Scheduler (see status.setter)
             self.status = new_status
 
-    def release_key(
+    def _purge_state(
         self,
         key: str,
-        cause: TaskState | None = None,
-        report: bool = True,
         *,
         stimulus_id: str,
     ) -> None:
-        try:
-            if self.validate:
-                assert not isinstance(key, TaskState)
-            ts = self.tasks[key]
-            # needed for legacy notification support
-            state_before = ts.state
-            ts.state = "released"
+        """Ensure that TaskState attributes are reset to a neutral default and
+        Worker-level state associated to the provided key is cleared (e.g.
+        who_has)
+        This is idempotent
+        """
+        ts = self.tasks[key]
+        logger.debug(
+            "Release key %s",
+            {"key": key, "stimulus_id": stimulus_id},
+        )
+        self.log.append((key, "purge-state", stimulus_id, time()))
+        if key in self.data:
+            del self.data[key]
+        if key in self.actors:
+            del self.actors[key]
 
-            logger.debug(
-                "Release key %s",
-                {"key": key, "cause": cause, "stimulus_id": stimulus_id},
-            )
-            if cause:
-                self.log.append(
-                    (key, "release-key", {"cause": cause}, stimulus_id, time())
-                )
-            else:
-                self.log.append((key, "release-key", stimulus_id, time()))
-            if key in self.data:
-                try:
-                    del self.data[key]
-                except FileNotFoundError:
-                    logger.error("Tried to delete %s but no file found", exc_info=True)
-            if key in self.actors:
-                del self.actors[key]
-
-            for worker in ts.who_has:
-                self.has_what[worker].discard(ts.key)
-            ts.who_has.clear()
-
-            if key in self.threads:
-                del self.threads[key]
-
-            if ts.resource_restrictions is not None:
-                if ts.state == "executing":
-                    for resource, quantity in ts.resource_restrictions.items():
-                        self.available_resources[resource] += quantity
-
-            for d in ts.dependencies:
-                ts.waiting_for_data.discard(d)
-                d.waiters.discard(ts)
-
-            ts.waiting_for_data.clear()
-            ts.nbytes = None
-            ts._previous = None
-            ts._next = None
-            ts.done = False
+        for worker in ts.who_has:
+            self.has_what[worker].discard(ts.key)
+        ts.who_has.clear()
 
-            self._executing.discard(ts)
-            self._in_flight_tasks.discard(ts)
+        if key in self.threads:
+            del self.threads[key]
```

Comment:
I typically prefer not to change something like this
- It is functionally equivalent but a different style. I prefer not changing style during functional refactoring to keep diffs easier to read unless there is a need to change something. In hindsight I shouldn't even have removed the try/except indentation in this PR
- This change in particular is already hard to read because git diff don't recognize the indentation. I find something like this much easier to review if there are identical code paths for me to use as anchor points.